File: LeetcodeJava/src/main/java/org/khwu/_301_remove_invlida_parentheses/LE301RemoveInvalidParentheses.java
package org.khwu._301_remove_invlida_parentheses;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class LE301RemoveInvalidParentheses {
    public List<String> removeInvalidParentheses(String s) {
        int n = s.length();
        char[] ar = s.toCharArray();
        int rmL = 0;
        int rmR = 0;
        for (int i = 0; i < n; i++) {
            if (ar[i] == '(') {
                rmL++;
            } else if (ar[i] == ')') {
                if (rmL == 0) {
                    rmR++;
                } else {
                    rmL--;
                }
            }
        }
        Set<String> res = new HashSet<>();
        dfs(ar, 0, new StringBuilder(), res, rmL, rmR, 0);

        return new ArrayList<>(res);
    }

    private void dfs(char[] ar, int i, StringBuilder sb, Set<String> res, int rmL, int rmR, int open) {
        if (rmL < 0 || rmR < 0 || open < 0) {
            return;
        }
        if (i == ar.length) {
            if (rmL == 0 && rmR == 0 && open == 0) res.add(sb.toString());
            return;
        }

        char c = ar[i];
        int n = sb.length();

        if (ar[i] == '(') {
            dfs(ar, i + 1, sb, res, rmL - 1, rmR, open);
            dfs(ar, i + 1, sb.append(c), res, rmL, rmR, open + 1);
        } else if (ar[i] == ')') {
            dfs(ar, i + 1, sb, res, rmL, rmR - 1, open);
            dfs(ar, i + 1, sb.append(c), res, rmL, rmR, open - 1);
        } else {
            dfs(ar, i + 1, sb.append(c), res, rmL, rmR, open);
        }

        sb.setLength(n);
    }
}

File: LeetcodePython/LE301RemoveInvalidParentheses.py
class Solution:
    def removeInvalidParentheses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        rmL, rmR = 0, 0
        for c in s:
            if c == '(':
                rmL += 1
            elif c == ')':
                if rmL:
                    rmL -= 1
                else:
                    rmR += 1

        res = set()

        def recursion(s, i, rmL, rmR, open, st):
            if rmL < 0 or rmR < 0 or open < 0:
                return
            if i == len(s):
                if rmL == 0 and rmR == 0 and open == 0:
                    res.add(str(st))
                return
            if s[i] == '(':
                recursion(s, i + 1, rmL - 1, rmR, open, st)
                recursion(s, i + 1, rmL, rmR, open + 1, st + s[i])
            elif s[i] == ')':
                recursion(s, i + 1, rmL, rmR - 1, open, st)
                recursion(s, i + 1, rmL, rmR, open - 1, st + s[i])
            else:
                recursion(s, i + 1, rmL, rmR, open, st + s[i])

        recursion(s, 0, rmL, rmR, 0, "")
        return list(res)